models/demos/falcon40b/tt/falcon_attention.py

```diff
@@ -382,69 +380,72 @@ def fwd_prefill(
             )
             key_layer[i].deallocate(True)
 
-        key_layer_transposed = convert_to_layout(
-            key_layer_transposed,
-            self.model_config["CREATE_KV_HEADS_OUTPUT_MEMCFG"],
-            self.model_config["DEFAULT_MEMCFG"],
-        )
-        attn_weights = []
-        for i in range(len(query_layer)):
-            attn_weights.append(
-                tt_lib.tensor.matmul(
-                    query_layer[i],
-                    key_layer_transposed[i],
-                    output_mem_config=self.model_config["PRE_SOFTMAX_MM_OUTPUT_MEMCFG"],
+        slice_size = 128
+        num_slices = q_len // slice_size
+
+        if num_slices > 1:
+            # Partially sliced and sharded attention
+            attn_output = []  # this is the output we write to. Initiate as empty tensors
+            for i in range(len(query_layer)):
+                attn_output.append(
+                    torch2tt_tensor(
+                        torch.zeros([1, self.num_heads_per_device, q_len, self.head_dim]),
+                        self.devices[i],
+                        tt_memory_config=self.model_config["DRAM_MEMCFG"],
+                        tt_dtype=self.model_config["POST_SOFTMAX_MM_OUTPUT_DTYPE"],
```

Comment:
Can we avoid this? In the decoder we will sum up attention output with residual and deallocate it - maybe we can skip deallocation and just pass it in to the next layer since the values inside don't matter?

I am not sure how much time we spend to push this tensor to device, if it is negligible then it probably doesn't make sense since it might not be intuitive passing in a tensor whose value is not used anywhere.

Reply:
Agree, could also move the initial allocation to model_preprocessing?

Reply:
Good point. The problem is, that the size of this depends on the sequence length, it can be different from inference to inference. So I don't think we can pre-allocate it. In theory we could reuse it for all layers, that's what @s-jovic meant, right? But before we sum it up we also AllGather over all devices, where we allocate a different memory area anyway, right? So...I guess we could allocate it the first time and then reuse, and not free it up after the AllGather? Do you think that makes sense in this case? I have no idea how long allocation/deallocation takes, I could measure and then see if it makes sense to avoid it?

Reply:
`model_preprocessing` already depends on the seq len, right? We call it for each user anyhow

Reply:
Sure, makes sense